### PR TITLE
remove iteratorFoldM fixes #1716

### LINKED ISSUE #1716: `Foldable` instance for `List` breaks substitution
The `Foldable` instance for `List` breaks substitution, as the following program demonstrates. The value of `result.value` differs between the first and subsequent calls.

```scala
import cats.instances.list._
import cats.Eval
import cats.syntax.foldable._

val result  = List.range(0,10).foldM(List.empty[Int])((accum, elt) => Eval.always(elt :: accum))

result.value
// res: List[Int] = List(9, 8, 7, 6, 5, 4, 3, 2, 1, 0)

result.value
// res: List[Int] = List(0)
```

I haven't had time to debug but I suspect it relates to the use of a mutable `ListBuffer` in the implementation of `tailRecM`.

### PR DESCRIPTION
**Pull request by kailuowang**
also related https://github.com/typelevel/cats/pull/1117/files#r123765689

### CODE REVIEW

File: tests/src/test/scala/cats/tests/FoldableTests.scala [hunk 1]
```diff
@@ -214,17 +209,23 @@ class FoldableTestsAdditional extends CatsSuite {
   }
 
   test(".foldLeftM short-circuiting optimality") {
-    // test that no more elements are evaluated than absolutely necessary
+    // test that no more than 1 elements are evaluated than absolutely necessary
```
Comment (peterneyens) [inline tests/src/test/scala/cats/tests/FoldableTests.scala]: Should we look if we can solve this additional evaluation with @TomasMikula's [suggestion](https://github.com/typelevel/cats/pull/1117/files#r123775020)?

Comment (kailuowang) [inline tests/src/test/scala/cats/tests/FoldableTests.scala]: updated.

**Comment by kailuowang** [comment]
weird, for some reason github doesn't allow me to request @TomasMikula to review.

File: tests/src/test/scala/cats/tests/FoldableTests.scala (hunk 1 above)
Comment (peterneyens) [inline tests/src/test/scala/cats/tests/FoldableTests.scala]: You forgot to update this comment.

**Comment by TomasMikula** [comment]
👍

However, it's still the case that most data structures can have a more efficient implementation of `foldM` than the default one from `Foldable`.

**Comment by kailuowang** [comment]
@TomasMikula yes, it's nice that you also mentioned it in the comment of `foldM`. we'll get to it in a later PR(s).

File: core/src/main/scala/cats/instances/list.scala [hunk 2]
```diff
@@ -90,9 +90,6 @@ trait ListInstances extends cats.kernel.instances.ListInstances {
 
       override def filter[A](fa: List[A])(f: A => Boolean): List[A] = fa.filter(f)
 
-      override def foldM[G[_], A, B](fa: List[A], z: B)(f: (B, A) => G[B])(implicit G: Monad[G]): G[B] =
-        Foldable.iteratorFoldM(fa.toIterator, z)(f)
```
Comment (johnynek) [inline core/src/main/scala/cats/instances/list.scala]: can't we maybe do this one?

```scala
def step(in: (List[A], B])): G[Either[(List[A], B), A]] = in match {
  case (Nil, b) => G.pure(Right(b))
  case (a :: tail, b) => G.map(f(a, b)) { bnext => Left((tail, bnext)) }
} 

G.tailRecM((fa, z))(step)
```
right? If we had this, we could cover the common case for `List` and `NonEmptyList`, no?

File: core/src/main/scala/cats/instances/stream.scala [hunk 3]
```diff
@@ -118,9 +118,6 @@ trait StreamInstances extends cats.kernel.instances.StreamInstances {
 
       override def collect[A, B](fa: Stream[A])(f: PartialFunction[A, B]): Stream[B] = fa.collect(f)
 
-      override def foldM[G[_], A, B](fa: Stream[A], z: B)(f: (B, A) => G[B])(implicit G: Monad[G]): G[B] =
-        Foldable.iteratorFoldM(fa.toIterator, z)(f)
```
Comment (johnynek) [inline core/src/main/scala/cats/instances/stream.scala]: we could copy the `List` approach above here:
```scala
def step(in: (Stream[A], B])): G[Either[(Stream[A], B), A]] = in match {
  case (Stream.empty, b) => G.pure(Right(b))
  case (a #:: tail, b) => G.map(f(a, b)) { bnext => Left((tail, bnext)) }
} 

G.tailRecM((fa, z))(step)
```

File: core/src/main/scala/cats/instances/vector.scala [hunk 4]
```diff
@@ -90,9 +90,6 @@ trait VectorInstances extends cats.kernel.instances.VectorInstances {
 
       override def collect[A, B](fa: Vector[A])(f: PartialFunction[A, B]): Vector[B] = fa.collect(f)
 
-      override def foldM[G[_], A, B](fa: Vector[A], z: B)(f: (B, A) => G[B])(implicit G: Monad[G]): G[B] =
-        Foldable.iteratorFoldM(fa.toIterator, z)(f)
```
Comment (johnynek) [inline core/src/main/scala/cats/instances/vector.scala]: `Vector.tail` is so slow, we are probably better off doing `fa.toList.foldM` here if we do the above.

File: core/src/main/scala/cats/instances/stream.scala (hunk 3 above)
Comment (kailuowang) [inline core/src/main/scala/cats/instances/stream.scala]: Patten match Stream into `a #:: tail` will force an evaluation of the head of the `tail`, so I used vanilla `if`

**Comment by kailuowang** [comment]
@johnynek thanks for helping me with these, it was too lazy of me.

File: core/src/main/scala/cats/instances/stream.scala [hunk 5]
```diff
@@ -118,8 +119,17 @@ trait StreamInstances extends cats.kernel.instances.StreamInstances {
...
+      override def foldM[G[_], A, B](fa: Stream[A], z: B)(f: (B, A) => G[B])(implicit G: Monad[G]): G[B] = {
```
Comment (TomasMikula) [inline core/src/main/scala/cats/instances/stream.scala]: This is good, but now again the default implementation of `foldM` in `Foldable` is not tested.

Comment (kailuowang) [inline core/src/main/scala/cats/instances/stream.scala]: oops. I'll correct the test.

Comment (kailuowang) [inline core/src/main/scala/cats/instances/stream.scala]: updated.

File: core/src/main/scala/cats/data/NonEmptyList.scala [hunk 6]
```diff
@@ -431,6 +431,9 @@ private[data] sealed trait NonEmptyListInstances extends NonEmptyListInstances0
       override def fold[A](fa: NonEmptyList[A])(implicit A: Monoid[A]): A =
         fa.reduce
 
+      override def foldM[G[_], A, B](fa: NonEmptyList[A], z: B)(f: (B, A) => G[B])(implicit G: Monad[G]): G[B] =
+        G.flatMap(f(z, fa.head))(bnext => Foldable[List].foldM(fa.tail, bnext)(f))
+
```
Comment (peterneyens) [inline core/src/main/scala/cats/data/NonEmptyList.scala]: `NonEmptyList` sort of gets this implementation already through `NonEmptyReducible`, but I guess this saves us the `split`.

File: core/src/main/scala/cats/instances/stream.scala (hunk 5 above)
Comment (peterneyens) [inline core/src/main/scala/cats/instances/stream.scala]: Here we would profit if had some machinery mentioned in #1684.

File: core/src/main/scala/cats/instances/stream.scala [hunk 7]
```diff
@@ -118,8 +119,17 @@ trait StreamInstances extends cats.kernel.instances.StreamInstances {
...
+        def step(in: (Stream[A], B)): G[Either[(Stream[A], B), B]] = {
+          val (s, b)  = in
+          if(s.isEmpty)
```
Comment (peterneyens) [inline core/src/main/scala/cats/instances/stream.scala]: Scalastyle seems to want a space between the `if` and the `(`.

File: core/src/main/scala/cats/data/NonEmptyList.scala (hunk 6 above)
Comment (kailuowang) [inline core/src/main/scala/cats/data/NonEmptyList.scala]: ah, forgot about that. A single split call probably doesn't justify the code duplication. I am going to remove this.

File: core/src/main/scala/cats/instances/stream.scala (hunk 5 above)
Comment (johnynek) [inline core/src/main/scala/cats/instances/stream.scala]: @TomasMikula I think we are still tested with `Map` and `Set` but I agree tests for overrides would be good.

Comment (kailuowang) [inline core/src/main/scala/cats/instances/stream.scala]: yup. I thought about that too.

**Comment by johnynek** [comment]
👍

**Comment by kailuowang** [comment]
@peterneyens any other issues/feedback?

**Pull request merged**